### PR TITLE
Update TCK process to address the issue 59

### PR DESCRIPTION
**Pull request by Emily-Jiang**

### CODE REVIEW

**Review by oliviergoulet5 — APPROVED** [review]

**Comment by Emily-Jiang** [comment]
@scottmarlow Can you please take a look at this PR to see whether you have any comments?

File: content/committees/specification/tckprocess/index.md
Comment (scottmarlow) [inline content/committees/specification/tckprocess/index.md:55]: This paragraph as written needs some changes.

Is your goal to ensure that the EE 11 Platform Specification does not go to ballot until all Type 1 tests are extended by a Type 2 test to allow that test to run as a Type 2 test?  Or is it more helpful to require that all Platform TCK testing is done with a Profile/Platform TCK?

I think we need to allow Type 1 tests to be the only tests for all Specification waves (exceptions are the profile/platform Specifications) as only the Type 1 tests can be passed without having an implementation of the higher Specification wave(s).

Including Type 2/3 tests in individual specification TCK introduces cycles between Specifications that cross the higher Specification waves.

**Comment by scottmarlow** [comment]
For reference issue link is https://github.com/jakartaee/specification-committee/issues/59

Some more context is on the various threads of the https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03128.html discussion.  

As per https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03134.html we should look at the technical debt more urgently (currently this means it is a blocking issue for the EE 11 release).  This is stated in the next message https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03135.html :-)

I also want to highlight https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03121.html from the email thread.  

The entire email conversation is worth reading though as it points out several problems that we have including the impact on the `development process` of waiting until later Specification waves to try using an individual spec in a profile or Platform Specification.  There are solutions to all of these problems, they just take time and need to be addressed in some order.  

One short term solution to discovering problems with early wave Specification very late during testing of a profile/Platform Specification wave is that the early wave Specification ballot needs to be restarted.  I'm getting off topic (of this pull request) with referencing the email discussion. :-)

File: content/committees/specification/tckprocess/index.md
Comment (Emily-Jiang) [reply content/committees/specification/tckprocess/index.md:55]: @scottmarlow The 2 types of tests are independent. Type 1 tests do not need to be extended by Type 2 tests. However, a spec must contain type 2/3 tests. This paragraph does not talk about cross spec tests. It is pretty much about how tests are executed whether it is under Java SE or Jakarta EE mode. Type 1 is Java SE tests while type 2/3 are Jakarta EE mode tests.

**Comment by Emily-Jiang** [comment]
> For reference issue link is [jakartaee/specification-committee#59](https://github.com/jakartaee/specification-committee/issues/59)
> 
> Some more context is on the various threads of the https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03128.html discussion.
> 
> As per https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03134.html we should look at the technical debt more urgently (currently this means it is a blocking issue for the EE 11 release). This is stated in the next message https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03135.html :-)
> 
> I also want to highlight https://www.eclipse.org/lists/jakarta.ee-spec.committee/msg03121.html from the email thread.
> 
> The entire email conversation is worth reading though as it points out several problems that we have including the impact on the `development process` of waiting until later Specification waves to try using an individual spec in a profile or Platform Specification. There are solutions to all of these problems, they just take time and need to be addressed in some order.
> 
> One short term solution to discovering problems with early wave Specification very late during testing of a profile/Platform Specification wave is that the early wave Specification ballot needs to be restarted. I'm getting off topic (of this pull request) with referencing the email discussion. :-)

@scottmarlow I think you mixed this intent with the cross spec tests. This clarification is to disallow the situation of a spec just containing Junit tests without any Arquillian style tests where you can deploy a .war and verify that app runs as speced.

**Comment by scottmarlow** [comment]
> @scottmarlow I think you mixed this intent with the cross spec tests. This clarification is to disallow the situation of a spec just containing Junit tests without any Arquillian style tests where you can deploy a .war and verify that app runs as speced.

Are you saying that as per the requirements defined in https://github.com/jakartaee/specification-committee/issues/59, that future specification releases like Jakarta Mail must provide Junit tests with a Arquillian style tests where you can deploy a .war and verify that app runs as specified?

**Comment by Emily-Jiang** [comment]
> > @scottmarlow I think you mixed this intent with the cross spec tests. This clarification is to disallow the situation of a spec just containing Junit tests without any Arquillian style tests where you can deploy a .war and verify that app runs as speced.
> 
> Are you saying that as per the requirements defined in [jakartaee/specification-committee#59](https://github.com/jakartaee/specification-committee/issues/59), that future specification releases like Jakarta Mail must provide Junit tests with a Arquillian style tests where you can deploy a .war and verify that app runs as specified?

yes

**Comment by scottmarlow** [comment]
> > > @scottmarlow I think you mixed this intent with the cross spec tests. This clarification is to disallow the situation of a spec just containing Junit tests without any Arquillian style tests where you can deploy a .war and verify that app runs as speced.
> > 
> > 
> > Are you saying that as per the requirements defined in [jakartaee/specification-committee#59](https://github.com/jakartaee/specification-committee/issues/59), that future specification releases like Jakarta Mail must provide Junit tests with a Arquillian style tests where you can deploy a .war and verify that app runs as specified?
> 
> yes

I propose that we avoid adding the Jakarta Mail Specification dependency on the Jakarta Platform and instead require that Jakarta EE Platform implementations will only need to pass TCK tests that can be run against a Platform (e.g. EE containers) implementation.  This would also help reduce the complexity of our development process.

**Comment by Emily-Jiang** [comment]
> > > > @scottmarlow I think you mixed this intent with the cross spec tests. This clarification is to disallow the situation of a spec just containing Junit tests without any Arquillian style tests where you can deploy a .war and verify that app runs as speced.
> > > 
> > > 
> > > Are you saying that as per the requirements defined in [jakartaee/specification-committee#59](https://github.com/jakartaee/specification-committee/issues/59), that future specification releases like Jakarta Mail must provide Junit tests with a Arquillian style tests where you can deploy a .war and verify that app runs as specified?
> > 
> > 
> > yes
> 
> I propose that we avoid adding the Jakarta Mail Specification dependency on the Jakarta Platform and instead require that Jakarta EE Platform implementations will only need to pass TCK tests that can be run against a Platform (e.g. EE containers) implementation. This would also help reduce the complexity of our development process.

That is the right approach.

**Comment by Emily-Jiang** [comment]
Based on today's spec committee call, this PR can be merged. However, I don't have a committer status to merge my PR. @paulbuck @ivargrimstad can you please merge this PR for me?

**Comment by scottmarlow** [comment]
Would we achieve the desired outcome with a more liberal requirement like:

> Jakarta EE Platform implementations only need to pass TCK tests that can be run against a Platform (e.g. EE containers) implementation to be considered Jakarta EE Platform compatible.  Jakarta EE Platform implementations do not need to run TCK tests that do not run against a Platform implementation.

**Comment by Emily-Jiang** [comment]
> Would we achieve the desired outcome with a more liberal requirement like:
> 
> > Jakarta EE Platform implementations only need to pass TCK tests that can be run against a Platform (e.g. EE containers) implementation to be considered Jakarta EE Platform compatible.  Jakarta EE Platform implementations do not need to run TCK tests that do not run against a Platform implementation.

No, not really. I am not seeking the clarification on how to certify implementation but ensure each specification has runnable TCKs that can be used to verify the implementations in the platform or profile environment.

**Comment by scottmarlow** [comment]
Could we reword to better reflect the intended requirements?  The current wording sounds like we are expecting the Mail/JSON-P/JSON-B Specification TCK to add Platform/Profile dependencies which (IMO) isn't an itch that Standalone Specifications need to scratch.  

Instead, we should require that future EE releases (including EE 11) validate all of the Platform/profile requirements via a Platform/profile level TCK, as was mentioned in the linked email thread.

**Comment by Emily-Jiang** [comment]
> Could we reword to better reflect the intended requirements? The current wording sounds like we are expecting the Mail/JSON-P/JSON-B Specification TCK to add Platform/Profile dependencies which (IMO) isn't an itch that Standalone Specifications need to scratch.
> 
> Instead, we should require that future EE releases (including EE 11) validate all of the Platform/profile requirements via a Platform/profile level TCK, as was mentioned in the linked email thread.

@scottmarlow feel free to suggest updates via a new PR. This is guideline to follow once it is merged. The past ones should be evaluated to see whether it is feasible to improve further.

**Pull request merged**
